### ISSUE TITLE
Approach to be taken in this library

### ISSUE THREAD

**Issue by dumblob** [tczajka/exact-real-rs#1]
I'm researching existing approaches of highly efficient implementations of exact arithmetic with reals - ideally with "infinite" precision ("infinite" as defined below).

My goal is actually to get a more high-school math feeling when using it - i.e. the final precision is solely based on the precision of the input.

This everything **without** any manual specification of the precision needed for any given arithmetic operation or function computation. Because real number literals already have the precision information - it's the number of digits. And e.g. binary floats received over network also have this information - it's the worst number of digits of all precisely representable base-10 numbers (there are usually four: two around 0 and two somewhere on the pathway to +-inf).

So far the closest is (quite surprisingly) [mpdecimal](https://www.bytereef.org/mpdecimal/ ) but it seems to have the major drawback that it doesn't support "dynamic" precision (with dynamic precision I mean `1.1 + 1.123 = 2.223` - i.e. both operands had different precision and assuming their representations were exact, we shall **automatically** get a result of at least the precision of the most precise one in case of the plus operation).

Could you elaborate on what are the plans here? Any links to research papers, existing libraries, etc.?

**Comment by tczajka** [comment]
The idea is that calculations are always in infinite precision -- not some finite precision based on the precision of inputs or anything like that. For instance, if you assign sqrt(2) to a variable, and then print it to a hundred digits, those hundred digits will be precise. If you then print it to a million digits, those will also be exact.

mpdecimal implements finite precision with rounding, so that's different.

**Comment by dumblob** [comment]
Perfect, now I understand.

Basically the actual evaluation happens in a deferred manner first at the point when the result is being queried for in which case it's mandatory to specify expected precision.

E.g. all expressions just construct a (potentially very large) graph and only after a "materialize() call" the graph would be used to get a result. The graph itself will stay in memory until the last possible "materialize() call" (thanks e.g. to Rust scoping, borrowing, etc.).

Or something similar to what [Herbie](https://herbie.uwplse.org/ ) and [Pareto Herbie](https://herbie.uwplse.org/arith21.html ) does.

Is that so?

**Comment by tczajka** [comment]
Yes that's the idea.

However there is no explicit `materialize` call. The library will internally compute with whatever precision is necessary to implement API calls. Logically, from the outside it should just look like a real number.

**Comment by dumblob** [comment]
Ok, very interesting. Will keep an eye on this. Any (preliminary) schedule?
